android/libraries/rib-router-navigator/src/main/kotlin/com/uber/rib/core/RouterAndState.kt

```diff
@@ -0,0 +1,153 @@
+   */
+  internal val router: R
+    get() = routerAccessor.router()
+
+  internal fun willAttachToHost(previousState: StateT?, isPush: Boolean) =
+    routerAccessor.safeOperation { safeRouter ->
+      attachTransition.willAttachToHost(safeRouter, previousState, state, isPush)
+    }
+
+  internal fun willDetachFromHost(newState: StateT?, isPush: Boolean) =
+    routerAccessor.safeOperation { safeRouter ->
+      detachCallback.willDetachFromHost(safeRouter, state, newState, isPush)
+    }
+
+  internal fun onPostDetachFromHost(newState: StateT?, isPush: Boolean) =
+    routerAccessor.safeOperation { safeRouter ->
+      detachCallback.onPostDetachFromHost(safeRouter, newState, isPush)
+    }
+
+  private val detachCallback: RouterNavigator.DetachCallback<R, StateT> by lazy {
+    RouterDestroyerCallbackWrapper(
+      baseCallback = wrapDetachTransitionIfNeed(detachTransition),
+      onDestroy = {
+        routerAccessor.safeConditionalDestroy(state) {
+          !forceRouterCaching && !state.isCacheable()
+        }
+      }
+    )
+  }
+
+  private fun wrapDetachTransitionIfNeed(
+    detachTransition: RouterNavigator.DetachTransition<R, StateT>?
+  ): RouterNavigator.DetachCallback<R, StateT>? {
+    return (detachTransition as? RouterNavigator.DetachCallback)
+      ?: detachTransition?.let { DetachCallbackWrapper(it) }
+  }
+
+  /**
+   * Wrapper class to wrap [transitionCallback] calls into the new [RouterNavigator.DetachCallback]
+   * format.
+   */
+  private inner class DetachCallbackWrapper(
+    private val transitionCallback: RouterNavigator.DetachTransition<R, StateT>
+  ) : RouterNavigator.DetachCallback<R, StateT>() {
+
+    override fun willDetachFromHost(
+      router: R,
+      previousState: StateT,
+      newState: StateT?,
+      isPush: Boolean
+    ) = transitionCallback.willDetachFromHost(router, previousState, newState, isPush)
+  }
+
+  /**
+   * Wrapper class to wrap [RouterNavigator.DetachCallback] and call [onDestroy] after
+   * [RouterNavigator.DetachCallback.onPostDetachFromHost]
+   */
+  private inner class RouterDestroyerCallbackWrapper(
+    private val baseCallback: RouterNavigator.DetachCallback<R, StateT>?,
+    private val onDestroy: () -> Unit
+  ) : RouterNavigator.DetachCallback<R, StateT>() {
+
+    override fun willDetachFromHost(
+      router: R,
+      previousState: StateT,
+      newState: StateT?,
+      isPush: Boolean
+    ) {
+      baseCallback?.willDetachFromHost(router, previousState, newState, isPush)
+    }
+
+    override fun onPostDetachFromHost(router: R, newState: StateT?, isPush: Boolean) {
+      baseCallback?.onPostDetachFromHost(router, newState, isPush)
+      onDestroy.invoke()
+    }
+  }
+
+  private class SafeRouterAccessor<R : Router<*>>(
+    private val routerBuilder: () -> R
+  ) {
+    private val lock = ReentrantLock()
+    private var _router: R? = null
+
+    fun router(): R = lock.withLock {
+      _router ?: routerBuilder.invoke().apply {
```

Comment:
It probably makes sense to use `.also { newRouter ->` instead of apply here, since we're not applying any new stuff to the newly create router (instead we're attaching it to `_router`)

Reply:
Make sense, I've used  `apply` because it's not requires to return any value from it but 'm agree the `let` is better for code readability